File: packages/provide-uterm-go/detection/extractor.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package detection

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// ExtractKV extracts structured key/value data from screen text using the
// configured patterns. kvConfig may be a single field config (map[string]any
// with a "field" key), a list of field configs ([]any of maps), or nil.
//
// It mirrors KVExtractor.extract with validation enabled (the engine calls it
// that way): the returned map includes a "_validation" entry unless nothing
// was extracted, in which case the result is nil.
func ExtractKV(screen string, kvConfig any) map[string]any {
	return kvExtract(screen, kvConfig, true)
}

// kvExtract is the full KVExtractor.extract with an explicit run_validation
// toggle (exposed to tests via ExtractKVWithValidation).
func kvExtract(screen string, kvConfig any, runValidation bool) map[string]any {
	if !pyTruthy(kvConfig) {
		return nil
	}
	configs, ok := normalizeConfigs(kvConfig)
	if !ok {
		return nil
	}

	extracted := map[string]any{}
	for _, cfg := range configs {
		name, value, matched := extractSingleField(screen, cfg)
		if matched {
			extracted[name] = value
		}
	}
	if len(extracted) == 0 {
		return nil
	}

	if runValidation {
		extracted["_validation"] = validateExtracted(extracted, configs)
	}
	return extracted
}

// ExtractKVWithValidation exposes the run_validation toggle for tests.
func ExtractKVWithValidation(screen string, kvConfig any, runValidation bool) map[string]any {
	return kvExtract(screen, kvConfig, runValidation)
}

// normalizeConfigs converts the polymorphic kv_config into a list of config
// maps, matching Python's single-dict / list handling. The bool reports
// whether the shape was valid.
func normalizeConfigs(kvConfig any) ([]map[string]any, bool) {
	switch cfg := kvConfig.(type) {
	case map[string]any:
		if _, hasField := cfg["field"]; hasField {
			return []map[string]any{cfg}, true
		}
		return nil, false
	case []map[string]any:
		return cfg, true
	case []any:
		out := make([]map[string]any, 0, len(cfg))
		for _, item := range cfg {
			if m, ok := item.(map[string]any); ok {
				out = append(out, m)
			}
		}
		return out, true
	default:
		return nil, false
	}
}

// extractSingleField extracts one field from screen using its config map,
// returning (field, value, matched). matched is false when the field/regex is
// missing, nothing matched, or the type conversion failed.
func extractSingleField(screen string, config map[string]any) (string, any, bool) {
	fieldName := patternString(config, "field", "")
	fieldType := patternString(config, "type", "string")
	pattern := patternString(config, "regex", "")

	if fieldName == "" || pattern == "" {
		return "", nil, false
	}

	re, err := compilePyRegex(pattern, "mi")
	if err != nil {
		return "", nil, false
	}
	// Use the last match: screen buffers carry scroll history, so the most
	// recent value is at the end.
	all := re.FindAllStringSubmatchIndex(screen, -1)
	if len(all) == 0 {
		return "", nil, false
	}
	loc := all[len(all)-1]

	var valueStr string
	if re.NumSubexp() >= 1 && loc[2] >= 0 { // group 1 participated
		valueStr = screen[loc[2]:loc[3]]
	} else {
		valueStr = screen[loc[0]:loc[1]]
	}

	value, ok := convertType(valueStr, fieldType)
	if !ok {
		return "", nil, false
	}
	return fieldName, value, true
}

// convertType converts a captured string to the target type, returning ok=false
// on conversion failure (mirroring Python's ValueError -> None).
func convertType(valueStr, targetType string) (any, bool) {
	valueStr = pyStrip(valueStr)

	switch targetType {
	case "string":
		return valueStr, true
	case "int":
		n, err := strconv.Atoi(strings.ReplaceAll(valueStr, ",", ""))
		if err != nil {
			return nil, false
		}
		return n, true
	case "float":
		f, err := strconv.ParseFloat(strings.ReplaceAll(valueStr, ",", ""), 64)
		if err != nil {
			return nil, false
		}
		return f, true
	case "bool":
		switch strings.ToLower(valueStr) {
		case "true", "yes", "y", "1", "on":
			return true, true
		case "false", "no", "n", "0", "off":
			return false, true
		default:
			return nil, false
		}
	default:
		return valueStr, true
	}
}

// validateExtracted validates extracted values against the config constraints,
// returning {"valid": bool, "errors": []string}. Error strings are byte-for-
// byte compatible with the Python implementation.
func validateExtracted(extracted map[string]any, configs []map[string]any) map[string]any {
	errors := []string{}
	for _, cfg := range configs {
		fieldRaw, hasField := cfg["field"]
		field, isStr := fieldRaw.(string)
		if !hasField || !isStr {
			continue
		}
		value, present := extracted[field]
		rules, _ := cfg["validate"].(map[string]any)
		required := pyTruthy(cfg["required"])
		fieldType := patternString(cfg, "type", "string")

		if required && !present {
			errors = append(errors, fmt.Sprintf("%s: required but not found", field))
			continue
		}
		if !present {
			continue
		}
		switch fieldType {
		case "int", "float":
			validateNumeric(field, value, rules, &errors, fieldType)
		case "string":
			validateString(field, value, rules, &errors)
		}
	}
	return map[string]any{"valid": len(errors) == 0, "errors": errors}
}

func validateNumeric(field string, value any, rules map[string]any, errors *[]string, fieldType string) {
	wantInt := fieldType == "int"
	typeOK := false
	switch value.(type) {
	case int, int64:
		typeOK = wantInt
	case float64:
		typeOK = !wantInt
	}
	if !typeOK {
		*errors = append(*errors, fmt.Sprintf("%s: expected %s, got %s", field, fieldType, pyTypeName(value)))
		return
	}
	valF, _ := toFloat(value)
	if mn, ok := rules["min"]; ok {
		if mnF, fok := toFloat(mn); fok && valF < mnF {
			*errors = append(*errors, fmt.Sprintf("%s: value %s below min %s", field, pyNum(value), pyNum(mn)))
		}
	}
	if mx, ok := rules["max"]; ok {
		if mxF, fok := toFloat(mx); fok && valF > mxF {
			*errors = append(*errors, fmt.Sprintf("%s: value %s exceeds max %s", field, pyNum(value), pyNum(mx)))
		}
	}
}

func validateString(field string, value any, rules map[string]any, errors *[]string) {
	s, ok := value.(string)
	if !ok {
		*errors = append(*errors, fmt.Sprintf("%s: expected string, got %s", field, pyTypeName(value)))
		return
	}
	if pat, ok := rules["pattern"].(string); ok {
		// Python re.match: the pattern must match starting at position 0
		// (no MULTILINE/IGNORECASE flags).
		re, err := compilePyRegex(pat, "")
		if err != nil || !matchAtStart(re, s) {
			*errors = append(*errors, fmt.Sprintf("%s: value '%s' does not match pattern %s", field, s, pat))
		}
	}
	if allowed, ok := rules["allowed_values"]; ok {
		if !inAllowed(s, allowed) {
			*errors = append(*errors, fmt.Sprintf("%s: value '%s' not in allowed values %s", field, s, pyList(allowed)))
		}
	}
}

// matchAtStart mirrors Python's re.match: the pattern must match starting at
// position 0 of s.
func matchAtStart(re *regexp.Regexp, s string) bool {
	loc := re.FindStringIndex(s)
	return loc != nil && loc[0] == 0
}

func inAllowed(value string, allowed any) bool {
	switch xs := allowed.(type) {
	case []any:
		for _, item := range xs {
			if s, ok := item.(string); ok && s == value {
				return true
			}
		}
	case []string:
		for _, item := range xs {
			if item == value {
				return true
			}
		}
	}
	return false
}

func pyTypeName(v any) string {
	switch v.(type) {
	case string:
		return "str"
	case int, int64:
		return "int"
	case float64:
		return "float"
	case bool:
		return "bool"
	default:
		return "NoneType"
	}
}

// pyNum formats an int/float the way Python's str() would inside an f-string.
func pyNum(v any) string {
	switch x := v.(type) {
	case int:
		return strconv.Itoa(x)
	case int64:
		return strconv.FormatInt(x, 10)
	case float64:
		return strconv.FormatFloat(x, 'g', -1, 64)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// pyList renders a slice the way Python's str(list) would (for allowed_values
// diagnostics), e.g. ['Normal', 'Test'].
func pyList(v any) string {
	items, ok := v.([]any)
	if !ok {
		return fmt.Sprintf("%v", v)
	}
	parts := make([]string, 0, len(items))
	for _, it := range items {
		if s, ok := it.(string); ok {
			parts = append(parts, "'"+s+"'")
		} else {
			parts = append(parts, pyNum(it))
		}
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

File: packages/provide-uterm-ts/src/detection/extractor.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Key-value extraction from screen text.
 *
 * Port of the Python module `provide.uterm.detection.extractor`.
 *
 * This turns a screen into the numbers an agent acts on, so every way it can
 * be quietly wrong ends with something acting on a stale or mistyped value.
 * A field that cannot be read is absent rather than zero, and validation is
 * reported alongside the values rather than enforced over them.
 */

import { compilePySearch, pyFloat, pyInt } from "../pycompat/index.ts";

/** One field to pull out of a screen. */
export interface ExtractConfig {
  field?: unknown;
  regex?: unknown;
  type?: unknown;
  validate?: unknown;
  required?: unknown;
  [key: string]: unknown;
}

/** What validation concluded. */
export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

/** Extracted values, with the validation report alongside them. */
export type Extracted = Record<string, unknown> & { _validation?: ValidationResult };

/** The words that mean yes. */
const TRUTHY: ReadonlySet<string> = new Set(["true", "yes", "y", "1", "on"]);

/** The words that mean no. */
const FALSY: ReadonlySet<string> = new Set(["false", "no", "n", "0", "off"]);

/**
 * Convert a matched string to the type a rule asked for.
 *
 * Returns nothing where the reference would raise, because a field that could
 * not be read has to be absent: a zero where the screen said "abc" is worse
 * than nothing, since a caller cannot tell it from a real zero.
 *
 * A type this does not know falls back to text, so a rule author still gets
 * their value.
 */
export function convertType(value: string, targetType: string): unknown {
  const text = value.trim();

  if (targetType === "int") {
    // Separators stripped first: screens write numbers for people to read.
    return pyInt(text.replaceAll(",", ""));
  }
  if (targetType === "float") {
    return pyFloat(text.replaceAll(",", ""));
  }
  if (targetType === "bool") {
    const lowered = text.toLowerCase();
    if (TRUTHY.has(lowered)) {
      return true;
    }
    if (FALSY.has(lowered)) {
      return false;
    }
    // "maybe" is not a decision, and guessing one is how an agent confirms
    // something it was never asked to confirm.
    return undefined;
  }
  return text;
}

/**
 * A field or pattern the config actually supplied.
 *
 * Tested for truthiness rather than for being a string, as the reference
 * does. A field name that is not a string is therefore kept — and skipped by
 * the validation pass, which does check — so such a value reaches a caller
 * unvalidated. Reproduced rather than tightened: a rule set may be relying on
 * the value being present at all.
 */
function supplied(value: unknown): unknown {
  return value === undefined || value === null || value === "" || value === 0 || value === false ? undefined : value;
}

/**
 * Pull one field out of a screen.
 *
 * The **last** match wins. A screen buffer holds scroll history, so the same
 * label appears many times and only the bottom one is current — reading the
 * first would take a credit balance from several screens ago.
 */
function extractSingleField(screen: string, config: ExtractConfig): [PropertyKey, unknown] | undefined {
  const field = supplied(config.field);
  const pattern = supplied(config.regex);
  if (field === undefined || pattern === undefined) {
    return undefined;
  }

  // Case-insensitive and per-line, as the reference compiles them: a screen's
  // case drifts between versions of a program, and a value can be anywhere.
  const compiled = new RegExp(compilePySearch(String(pattern), { ignoreCase: true }).source, "gmi");
  let last: RegExpExecArray | undefined;
  for (const match of screen.matchAll(compiled)) {
    last = match;
  }
  if (last === undefined) {
    return undefined;
  }

  // A capture group is the value when there is one, so a pattern written with
  // parentheses yields the number rather than the label and the number.
  const text = last[1] ?? last[0];
  const converted = convertType(text, typeof config.type === "string" ? config.type : "string");
  return converted === undefined ? undefined : [field as PropertyKey, converted];
}

/** Check one numeric field. */
function validateNumeric(
  field: string,
  value: unknown,
  rules: Record<string, unknown>,
  errors: string[],
  fieldType: string,
): void {
  // A float field accepts a whole number: the reference converted it with
  // `float()`, so 2 is a float there however it prints. Only an int field
  // needs the value to be whole.
  //
  // A boolean passes the int check and fails the float one, because a bool
  // *is* an int in the reference's type system and is not a float. Two
  // configs may name one field, so a value converted as a bool can be
  // validated as a number — refusing it would report an error the reference
  // does not.
  const isNumber = typeof value === "number";
  const acceptable =
    fieldType === "int"
      ? typeof value === "boolean" || (isNumber && Number.isInteger(value))
      : isNumber;
  if (!acceptable) {
    errors.push(`${field}: expected ${fieldType}, got ${describeType(value)}`);
    return;
  }
  const numeric = Number(value);
  if (typeof rules.min === "number" && numeric < rules.min) {
    errors.push(`${field}: value ${numeric} below min ${rules.min}`);
  }
  if (typeof rules.max === "number" && numeric > rules.max) {
    errors.push(`${field}: value ${numeric} exceeds max ${rules.max}`);
  }
}

/** Check one string field. */
function validateString(field: string, value: unknown, rules: Record<string, unknown>, errors: string[]): void {
  if (typeof value !== "string") {
    errors.push(`${field}: expected string, got ${describeType(value)}`);
    return;
  }
  // Anchored at the start and not at the end, because the reference matches
  // rather than fullmatches: "^A" means "begins with A".
  if (typeof rules.pattern === "string" && !new RegExp(`^(?:${compilePySearch(rules.pattern).source})`).test(value)) {
    errors.push(`${field}: value '${value}' does not match pattern ${rules.pattern}`);
  }
  if (Array.isArray(rules.allowed_values) && !rules.allowed_values.includes(value)) {
    errors.push(`${field}: value '${value}' not in allowed values ${pyRepr(rules.allowed_values)}`);
  }
}

/**
 * A list as the reference prints it.
 *
 * The message is what an operator reads, and it is compared against the
 * reference's output — Python renders a list of strings with single quotes
 * and a space after each comma, which JSON does not.
 */
function pyRepr(value: readonly unknown[]): string {
  return `[${value.map((item) => (typeof item === "string" ? `'${item}'` : String(item))).join(", ")}]`;
}

/** The reference's name for a value's type, for an error message. */
function describeType(value: unknown): string {
  if (typeof value === "boolean") {
    return "bool";
  }
  if (typeof value === "number") {
    return Number.isInteger(value) ? "int" : "float";
  }
  // Only ever a string by here: the conversions produce a string, a number or
  // a boolean, and the other two are named above.
  return "str";
}

/** Check every extracted value against the rules that produced it. */
function validate(extracted: Record<string, unknown>, configs: ExtractConfig[]): ValidationResult {
  const errors: string[] = [];
  for (const config of configs) {
    const field = config.field;
    // A field name that is not a string is skipped here but not during
    // extraction, so such a value reaches a caller unvalidated. Reproduced
    // because a rule set may be relying on the value being present. It
    // happens to reach the same result either way in this language, where an
    // object index is a string regardless — but the reference's dict is not
    // so forgiving, and the skip is what makes the two agree.
    if (typeof field !== "string") {
      continue;
    }
    const value = extracted[field];
    const rules = (config.validate ?? {}) as Record<string, unknown>;
    const fieldType = typeof config.type === "string" ? config.type : "string";

    // Truthiness, as the reference reads it: a rules file writing `1` means
    // required, and treating it as optional would let a missing field pass
    // silently.
    if (config.required && value === undefined) {
      errors.push(`${field}: required but not found`);
      continue;
    }
    if (value === undefined) {
      continue;
    }
    if (fieldType === "int" || fieldType === "float") {
      validateNumeric(field, value, rules, errors, fieldType);
    } else if (fieldType === "string") {
      validateString(field, value, rules, errors);
    }
  }
  // Every problem, not just the first: an operator fixing a rules file wants
  // the whole list rather than one error per run.
  return { valid: errors.length === 0, errors };
}

/**
 * Pull configured values out of a screen.
 *
 * Returns nothing when there was no usable configuration or nothing matched —
 * an empty result and no result are different, and the second tells a caller
 * the screen was not the one they expected.
 *
 * Validation is reported alongside the values rather than enforced over them,
 * so a caller can act on one it knows is out of range, or refuse to. Dropping
 * it would leave them unable to tell "absent" from "implausible".
 */
export function extractKV(
  screen: string,
  config: ExtractConfig | ExtractConfig[] | null | undefined,
  runValidation = true,
): Extracted | undefined {
  if (config === null || config === undefined) {
    return undefined;
  }

  let configs: ExtractConfig[];
  if (Array.isArray(config)) {
    configs = config;
  } else if (typeof config === "object" && Object.hasOwn(config, "field")) {
    configs = [config];
  } else {
    // A lone object with no `field` is not a configuration. The per-field
    // check below would reject it anyway; refusing here says that a bare
    // object was never one, rather than that it produced nothing.
    return undefined;
  }
  if (configs.length === 0) {
    // Likewise: no configurations and no matches both end in nothing, and
    // this says which happened.
    return undefined;
  }

  const extracted: Record<string, unknown> = {};
  for (const entry of configs) {
    const result = extractSingleField(screen, entry);
    if (result !== undefined) {
      extracted[String(result[0])] = result[1];
    }
  }

  if (Object.keys(extracted).length === 0) {
    return undefined;
  }
  if (runValidation) {
    extracted._validation = validate(extracted, configs);
  }
  return extracted;
}
